In [ ]:
# Import necessary libraries
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix


In [ ]:
# Load the Iris dataset
iris = datasets.load_iris()

In [ ]:
df = pd.DataFrame(data = iris.data, columns = iris.feature_names)

In [ ]:
df

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2
...,...,...,...,...
145,6.7,3.0,5.2,2.3
146,6.3,2.5,5.0,1.9
147,6.5,3.0,5.2,2.0
148,6.2,3.4,5.4,2.3


In [ ]:
df['target'] = iris.target
df['target_name'] = iris.target_names[iris.target]
df

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),target,target_name
0,5.1,3.5,1.4,0.2,0,setosa
1,4.9,3.0,1.4,0.2,0,setosa
2,4.7,3.2,1.3,0.2,0,setosa
3,4.6,3.1,1.5,0.2,0,setosa
4,5.0,3.6,1.4,0.2,0,setosa
...,...,...,...,...,...,...
145,6.7,3.0,5.2,2.3,2,virginica
146,6.3,2.5,5.0,1.9,2,virginica
147,6.5,3.0,5.2,2.0,2,virginica
148,6.2,3.4,5.4,2.3,2,virginica


In [ ]:
# Independent features (X) and dependent variable (y)
X = iris.data
y = iris.target

In [ ]:
# Train-test split (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [ ]:
# Create and train the SVM classifier
svm_model = SVC(kernel='rbf', C=1.0, gamma='scale')
svm_model.fit(X_train, y_train)


SVC()

Setosa is linearly separable.

Versicolor and Virginica overlap slightly — not fully separable by a straight line.
Hence, the RBF kernel allows the SVM to form a curved decision boundary that correctly separates these two species.

In [ ]:
y_pred = svm_model.predict(X_test)


In [ ]:
print("Training Accuracy:", svm_model.score(X_train, y_train))
print("Test Accuracy:", svm_model.score(X_test, y_test))


Training Accuracy: 0.9666666666666667
Test Accuracy: 1.0


Your training accuracy (96.6%) and test accuracy (100%) are very close.

This means the model generalizes well — it performs equally well on unseen data.

The small difference (3.4%) is normal and actually a good sign — it shows the model isn’t memorizing the data.

In [ ]:
from sklearn.model_selection import cross_val_score

scores = cross_val_score(svm_model, X, y, cv=5)
print("Cross-validation scores:", scores)
print("Average accuracy:", scores.mean())

Cross-validation scores: [0.96666667 0.96666667 0.96666667 0.93333333 1.        ]
Average accuracy: 0.9666666666666666


Cross-validation tests the model on different subsets of data multiple times.
If the accuracy stays consistent:

The model has learned general patterns rather than memorizing training data.

In your case, the model’s mean accuracy (≈96.7%) matches both your training (96.6%) and test (100%) accuracy → perfect alignment ✅

In [ ]:
# Accuracy
print("Accuracy:", accuracy_score(y_test, y_pred))


Accuracy: 1.0


In [ ]:
# Confusion Matrix
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))


Confusion Matrix:
 [[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]


In [ ]:
# Classification Report
print("\nClassification Report:\n", classification_report(y_test, y_pred, target_names=iris.target_names))


Classification Report:
               precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        10
  versicolor       1.00      1.00      1.00         9
   virginica       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30



In [ ]:
# Create a DataFrame for test results
test_results = pd.DataFrame(scaler.inverse_transform(X_test), columns=iris.feature_names)
test_results['actual_flower'] = iris.target_names[y_test]
test_results['predicted_flower'] = iris.target_names[y_pred]

# View rows
print(test_results)

    sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 6.1               2.8                4.7               1.2   
1                 5.7               3.8                1.7               0.3   
2                 7.7               2.6                6.9               2.3   
3                 6.0               2.9                4.5               1.5   
4                 6.8               2.8                4.8               1.4   
5                 5.4               3.4                1.5               0.4   
6                 5.6               2.9                3.6               1.3   
7                 6.9               3.1                5.1               2.3   
8                 6.2               2.2                4.5               1.5   
9                 5.8               2.7                3.9               1.2   
10                6.5               3.2                5.1               2.0   
11                4.8               3.0 

In [ ]:
# Save model and scaler
import joblib
joblib.dump(svm_model, 'svm_model.pkl')
joblib.dump(scaler, 'scaler.pkl')

['scaler.pkl']

In [ ]:
import joblib

joblib.dump(log_reg, "svm_model.pkl")
joblib.dump(scaler, "scaler.pkl")

print("Model and scaler saved successfully!")